src/zks_wallet/wallet.rs

```diff
@@ -255,28 +281,191 @@ where
-            .request_l2_transaction(
-                to,
-                l2_value,
-                call_data,
+    async fn deposit_erc20_token(
+        &self,
+        l1_token_address: Address,
+        amount: U256,
+        to: Address,
+        operator_tip: U256,
+        bridge_address: Option<Address>,
+        max_fee_per_gas: Option<U256>,
+        gas_price: Option<U256>,
+    ) -> Result<TransactionReceipt, ZKSWalletError<M, D>>
+    where
+        M: ZKSProvider,
+    {
+        let eth_provider = self.get_eth_provider()?;
+        let era_provider = self.get_era_provider()?;
+
+        let gas_limit: U256 = {
+            let address_str = format!("{l1_token_address:?}");
+            let default_erc20_deposit_gas_limit = 300000_u64; // FIXME make it a constant.
+            let is_mainnet = self.get_era_provider()?.get_chainid().await? == 324_i32.into();
+            if is_mainnet {
+                (*ERC20_DEPOSIT_GAS_LIMITS)
+                    .get(&address_str)
+                    .unwrap_or(&default_erc20_deposit_gas_limit)
+                    .to_owned() // FIXME fix unwrap
+            } else {
+                default_erc20_deposit_gas_limit
+            }
+        }
+        .into();
+
+        // If the user has already provided max_fee_per_gas or gas_price, we will use
+        // it to calculate the base cost for the transaction
+        let gas_price = if let Some(max_fee_per_gas) = max_fee_per_gas {
+            max_fee_per_gas
+        } else if let Some(gas_price) = gas_price {
+            gas_price
+        } else {
+            era_provider.get_gas_price().await?
+        };
+
+        let l2_gas_limit = U256::from(3_000_000_u32);
+
+        let base_cost: U256 = self
+            .get_base_cost(
                 l2_gas_limit,
-                gas_per_pubdata_byte,
-                Default::default(),
-                refund_recipient,
+                DEPOSIT_GAS_PER_PUBDATA_LIMIT.into(),
                 gas_price,
-                gas_limit,
-                l1_value,
             )
             .await?;
 
-        Ok(receipt)
+        // ERC20 token, `msg.value` is used only for the fee.
+        let value = base_cost + operator_tip;
+
+        let data: Bytes = {
+            let bridge_contract = abi::l1_bridge_contract();
+
+            #[allow(clippy::expect_used)]
+            let contract_function = bridge_contract
+                .function("deposit")
+                .expect("failed to get deposit function parameters");
+
+            let params = (
+                to,
+                l1_token_address,
+                amount,
+                l2_gas_limit,
+                U256::from(DEPOSIT_GAS_PER_PUBDATA_LIMIT),
+            );
+
+            #[allow(clippy::expect_used)]
+            contract_function
+                .encode_input(&params.into_tokens())
+                .expect("failed to encode deposit function parameters")
+                .into()
+        };
```

Comment:
We should avoid using `expect` here. The function already returns a `Result` so bubble up the error might be the best option

Reply:
I decided to use `expect` in this case, because the error state should be unreachable. That's why I consciously included `#[allow(clippy::expect_used)]`.

In this case, bubbling up this error would force the function user to take into account an unreachable type of error.

Reply:
Created issue https://github.com/lambdaclass/zksync-web3-rs/issues/103 to track this one